File: power-adapters-sample/src/main/java/com/nextfaze/poweradapters/sample/FileTree.java
package com.nextfaze.poweradapters.sample;

import android.support.annotation.Nullable;
import com.nextfaze.poweradapters.Condition;
import com.nextfaze.poweradapters.Holder;
import com.nextfaze.poweradapters.PowerAdapter;
import com.nextfaze.poweradapters.TreeAdapter;
import com.nextfaze.poweradapters.binding.AbstractBinder;
import com.nextfaze.poweradapters.binding.Binder;
import com.nextfaze.poweradapters.binding.ListBindingAdapter;
import com.nextfaze.poweradapters.data.Data;
import com.nextfaze.poweradapters.data.DataBindingAdapter;
import lombok.NonNull;

import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

import static com.nextfaze.poweradapters.PowerAdapter.concat;
import static com.nextfaze.poweradapters.sample.Utils.emptyMessage;
import static com.nextfaze.poweradapters.sample.Utils.loadingIndicator;
import static java.util.Collections.singletonList;

final class FileTree {

    private static final Object VIEW_TYPE_FILE = new Object();
    private static final Object VIEW_TYPE_PEEK = new Object();

    @NonNull
    static PowerAdapter createAdapter(@Nullable Data<File> overrideData, @NonNull File file) {
        return createAdapter(overrideData, file, new Tree(), 0);
    }

    @NonNull
    private static PowerAdapter createAdapter(@Nullable Data<File> overrideData,
                                              @NonNull File file,
                                              @NonNull Tree tree,
                                              int depth) {
        // Use the supplied Data if specified, otherwise create Data for directory contents.
        Data<File> data = overrideData != null ? overrideData : new DirectoryData(file);

        // Binder for a file/directory item.
        Binder<File, FileView> binder = new AbstractBinder<File, FileView>(R.layout.file_tree_file_item) {
            @Override
            public void bindView(@NonNull File file, @NonNull FileView v, @NonNull Holder holder) {
                v.setFile(file);
                v.setDepth(depth);
                v.setClickable(file.isDirectory());
                v.setOnClickListener(file.isDirectory() ? v1 -> tree.toggle(file, Flag.EXPANDED) : null);
                v.setOnPeekListener(file.isDirectory() ? () -> tree.togglePeek(file) : null);
            }

            @NonNull
            @Override
            public Object getViewType(@NonNull File file, int position) {
                return VIEW_TYPE_FILE;
            }
        };

        PowerAdapter adapter = new DataBindingAdapter(data, binder)
                .compose(nest(position -> {
                    File f = data.get(position);
                    PowerAdapter childAdapter = createPeekAdapter(f, tree);
                    if (f.isDirectory()) {
                        PowerAdapter childFilesAdapter = createAdapter(null, f, tree, depth + 1)
                                .showOnlyWhile(tree.isSet(f, Flag.EXPANDED));
                        return childAdapter.append(childFilesAdapter);
                    } else {
                        return childAdapter;
                    }
                }));

        // Add loading indicator and empty message.
        return concat(adapter, loadingIndicator(data), emptyMessage(data));
    }

    @NonNull
    private static PowerAdapter createPeekAdapter(@NonNull File file, @NonNull Tree tree) {
        Binder<File, FilePeekView> peekBinder = new AbstractBinder<File, FilePeekView>(R.layout.file_tree_peek_item) {
            @Override
            public void bindView(@NonNull File file, @NonNull FilePeekView v, @NonNull Holder holder) {
                v.setFile(file);
            }

            @NonNull
            @Override
            public Object getViewType(@NonNull File file, int position) {
                return VIEW_TYPE_PEEK;
            }
        };
        return new ListBindingAdapter<>(peekBinder, singletonList(file)).showOnlyWhile(tree.isSet(file, Flag.PEEKING));
    }

    @NonNull
    private static PowerAdapter.Transformer nest(@NonNull TreeAdapter.ChildAdapterSupplier childAdapterSupplier) {
        return adapter -> {
            TreeAdapter treeAdapter = new TreeAdapter(adapter, childAdapterSupplier);
            treeAdapter.setAutoExpand(true);
            return treeAdapter;
        };
    }

    /** Contains tree metadata, which controls expansion/peeking. */
    private static final class Tree {

        @NonNull
        private final Map<File, EnumSet<Flag>> mFlags = new HashMap<>();

        @NonNull
        private final Set<FlagCondition> mConditions = new CopyOnWriteArraySet<>();

        void togglePeek(@NonNull File file) {
            for (Map.Entry<File, EnumSet<Flag>> entry : mFlags.entrySet()) {
                if (!entry.getKey().equals(file)) {
                    entry.getValue().remove(Flag.PEEKING);
                }
            }
            EnumSet<Flag> flags = flags(file);
            if (flags.contains(Flag.PEEKING)) {
                flags.remove(Flag.PEEKING);
            } else {
                flags.add(Flag.PEEKING);
            }
            dispatchChanged();
        }

        void toggle(@NonNull File file, @NonNull Flag flag) {
            EnumSet<Flag> flags = flags(file);
            if (flags.contains(flag)) {
                if (flags.remove(flag)) {
                    dispatchChanged();
                }
            } else {
                if (flags.add(flag)) {
                    dispatchChanged();
                }
            }
        }

        @NonNull
        private EnumSet<Flag> flags(@NonNull File file) {
            EnumSet<Flag> flags = mFlags.get(file);
            if (flags == null) {
                flags = EnumSet.noneOf(Flag.class);
                mFlags.put(file, flags);
            }
            return flags;
        }

        private void dispatchChanged() {
            for (FlagCondition condition : mConditions) {
                condition.notifyChangedInternal();
            }
        }

        @NonNull
        Condition isSet(@NonNull File file, @NonNull Flag flag) {
            return new FlagCondition() {
                @Override
                public boolean eval() {
                    Set<Flag> flags = mFlags.get(file);
                    return flags != null && flags.contains(flag);
                }
            };
        }

        private abstract class FlagCondition extends Condition {
            @Override
            protected void onFirstObserverRegistered() {
                mConditions.add(this);
            }

            @Override
            protected void onLastObserverUnregistered() {
                mConditions.remove(this);
            }

            void notifyChangedInternal() {
                notifyChanged();
            }
        }
    }

    /** Enum of possible flags each {@link File} in a {@link Tree} can have. */
    private enum Flag {
        EXPANDED, PEEKING
    }
}

File: power-adapters-sample/src/main/java/com/nextfaze/poweradapters/sample/files/FileTree.kt
package com.nextfaze.poweradapters.sample.files

import android.arch.lifecycle.ViewModel
import com.jakewharton.rx.replayingShare
import com.nextfaze.poweradapters.PowerAdapter
import com.nextfaze.poweradapters.adapter
import com.nextfaze.poweradapters.binder
import com.nextfaze.poweradapters.data.rxjava2.ObservableDataBuilder
import com.nextfaze.poweradapters.data.toAdapter
import com.nextfaze.poweradapters.rxjava2.showOnlyWhile
import com.nextfaze.poweradapters.sample.R
import com.nextfaze.poweradapters.sample.emptyMessage
import com.nextfaze.poweradapters.sample.files.FileTreeViewModel.Flag.EXPANDED
import com.nextfaze.poweradapters.sample.files.FileTreeViewModel.Flag.PEEKING
import com.nextfaze.poweradapters.sample.loadingIndicator
import com.nextfaze.poweradapters.sample.nest
import com.nextfaze.poweradapters.sample.toggle
import com.nextfaze.poweradapters.toAdapter
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import kotlinx.collections.immutable.ImmutableMap
import kotlinx.collections.immutable.ImmutableSet
import kotlinx.collections.immutable.immutableMapOf
import kotlinx.collections.immutable.immutableSetOf
import kotlinx.collections.immutable.mutate
import kotlinx.collections.immutable.toImmutableSet

class FileTreeViewModel : ViewModel() {
    /** Root directory to present files from. */
    val root = File.rootDir()

    /** Maintains the set of [Flag]s for each [File]. */
    private val fileToFlagSetSubject =
            BehaviorSubject.createDefault<ImmutableMap<File, ImmutableSet<Flag>>>(immutableMapOf())

    /** Effectively caches directory content listing. */
    private val dirContentsCache = mutableMapOf<File, Observable<List<File>>>()

    /** Enables/disables [PEEKING] for [file], also clearing it from other files. */
    fun togglePeeking(file: File) {
        // Stop peeking at all other files
        mutate { fileToFlagSet ->
            fileToFlagSet.entries.forEach {
                if (it.key != file) {
                    it.setValue(it.value.filter { it != PEEKING }.toImmutableSet())
                }
            }
        }
        // Toggle peeking at this file
        toggle(file, PEEKING)
    }

    /** Enables/disables [flag] for [file]. */
    fun toggle(file: File, flag: Flag) = mutate {
        it[file] = it.getOrPut(file) { immutableSetOf() }.mutate { it.toggle(flag) }
    }

    /** Emits if [file] has [flag] currently enabled. */
    fun isSet(file: File, flag: Flag): Observable<Boolean> = fileToFlagSetSubject.map { flag in it[file].orEmpty() }

    /** Contents of a directory. */
    fun files(file: File): Observable<List<File>> = dirContentsCache.getOrPut(file) { file.files().replayingShare() }

    private fun mutate(body: (MutableMap<File, ImmutableSet<Flag>>) -> Unit) =
            fileToFlagSetSubject.onNext(fileToFlagSetSubject.value!!.mutate(body))

    /** Enum of possible flags each [File] in a [FileTree] can have.  */
    enum class Flag { EXPANDED, PEEKING }
}

/** Creates an adapter that presents [viewModel]. */
fun createFileTreeAdapter(viewModel: FileTreeViewModel) = createFileTreeAdapter(viewModel.root, viewModel, 0)

private fun createFileTreeAdapter(
        file: File,
        viewModel: FileTreeViewModel,
        depth: Int
): PowerAdapter = adapter {
    // Binds a file to a view
    val fileBinder = binder<File, FileView>(R.layout.file_tree_file_item) { container, f, holder ->
        setFile(f)
        setDepth(depth)
        isClickable = f.isDirectory
        setOnClickListener {
            if (f.isDirectory) viewModel.toggle(f, EXPANDED)
            container.scrollToPosition(holder.position)
        }
        setOnPeekListener { if (f.isDirectory) viewModel.togglePeeking(f) }
    }

    // File listing
    val files = viewModel.files(file).share()
    val data = ObservableDataBuilder<File>()
            .contents(files)
            // Compare attributes that determine identity
            .identityEquality { a, b -> a.name == b.name }
            // Compare attributes that affect presentation
            .contentEquality { a, b -> a.name == b.name && a.size == b.size }
            .build()
    +data.toAdapter(fileBinder).nest {
        adapter {
            val f = data[it]
            // Single-item peek adapter
            +peekAdapter(f, viewModel)
            // If directory, the file adapter representing the contents
            if (f.isDirectory) {
                +createFileTreeAdapter(f, viewModel, depth + 1).showOnlyWhile(viewModel.isSet(f, EXPANDED))
            }
        }
    }

    // Loading indicator
    +loadingIndicator(files)

    // Empty message
    +emptyMessage(files)
}

/** Creates an adapter that presents a (fake) preview row for the contents of [file]. */
private fun peekAdapter(file: File, viewModel: FileTreeViewModel): PowerAdapter {
    val peekBinder = binder<File, FilePeekView>(R.layout.file_tree_peek_item) { _, f, _ -> setFile(f) }
    return listOf(file).toAdapter(peekBinder).showOnlyWhile(viewModel.isSet(file, PEEKING))
}


